### PR TITLE
Make AttributeAccessInterface and CommandHandlerInterface registries consistent as objects with singletons

### PR DESCRIPTION
**Pull request by andy31415**
This makes both AAI and CHI consistent: objects with an `Instance` method and UpperCamelCase method names.

This is a followup on #34414

### CODE REVIEW

File: src/app/InteractionModelEngine.cpp [hunk 1]
```diff
@@ -99,11 +99,6 @@ CHIP_ERROR InteractionModelEngine::Init(Messaging::ExchangeManager * apExchangeM
 void InteractionModelEngine::Shutdown()
 {
     mpExchangeMgr->GetSessionManager()->SystemLayer()->CancelTimer(ResumeSubscriptionsTimerCallback, this);
-
-    // TODO: individual object clears the entire command handler interface registry.
-    //       This may not be expected.
-    CommandHandlerInterfaceRegistry::UnregisterAllHandlers();
```
Comment (bzbarsky-apple) [inline src/app/InteractionModelEngine.cpp:105]: Well, hold on.  We do in fact need to UnregisterAllHandlers() somewhere on shutdown.  Where is that happening?

File: src/app/clusters/resource-monitoring-server/resource-monitoring-server.h [hunk 2]
```diff
@@ -74,7 +74,8 @@ class Instance : public CommandHandlerInterface, public AttributeAccessInterface
      * @die                                     If the endpoint and cluster ID have not been enabled in zap.
      * @return CHIP_ERROR_INVALID_ARGUMENT      If the CommandHandler or Attribute Handler could not be registered.
      * @return CHIP_ERROR_INCORRECT_STATE       If the CommandHandler was already registered
-     * @return CHIP_ERROR_INCORRECT_STATE       If the registerAttributeAccessOverride fails.
+     * @return CHIP_ERROR_INCORRECT_STATE       If the
+     * AttributeAccessInterfaceRegistry::Instance().RegisterAttributeAccessOverride fails.
```
Comment (bzbarsky-apple) [inline src/app/clusters/resource-monitoring-server/resource-monitoring-server.h:78]: Weird indentation here... should have some spaces, no?

File: src/app/AttributeAccessInterfaceRegistry.h [hunk 3]
```diff
@@ -16,39 +16,50 @@
...
+     * endpoint it's registered for is disabled (or until shutdown if it's
+     * registered for all endpoints) or until it is explicitly unregistered.
+     * Registration will fail if there is an already-registered override for the
+     * same set of attributes.
+     *
+     * @return false if there is an existing override that the new one would
+     *               conflict with.  In this case the override is not registered.
+     * @return true if registration was successful.
+     */
+    bool RegisterAttributeAccessOverride(chip::app::AttributeAccessInterface * attrOverride);
```
Comment (bzbarsky-apple) [inline src/app/AttributeAccessInterfaceRegistry.h:39]: I know you kept the existing naming, but that naming was for a context-less free function.  In the context of this class, should this be perhaps called `Register`?  Or `RegisterAttributeAccess` or something?

Comment (bzbarsky-apple) [inline src/app/AttributeAccessInterfaceRegistry.h:39]:
```suggestion
    bool RegisterAttributeAccessOverride(AttributeAccessInterface * attrOverride);
```

File: src/app/AttributeAccessInterfaceRegistry.h [hunk 4]
```diff
@@ -16,39 +16,50 @@
...
 
-/**
- * Unregister all attribute access interfaces that match this given endpoint.
- */
-void unregisterAllAttributeAccessOverridesForEndpoint(EmberAfDefinedEndpoint * definedEndpoint);
+    /**
+     * Unregister an attribute access override (for example if the object
+     * implementing AttributeAccessInterface is being destroyed).
+     */
+    void UnregisterAttributeAccessOverride(chip::app::AttributeAccessInterface * attrOverride);
```
Comment (bzbarsky-apple) [inline src/app/AttributeAccessInterfaceRegistry.h:45]:
```suggestion
    void UnregisterAttributeAccessOverride(AttributeAccessInterface * attrOverride);
```

Comment (bzbarsky-apple) [inline src/app/AttributeAccessInterfaceRegistry.h:45]: Again, perhaps just `Unregister`?

File: src/app/AttributeAccessInterfaceRegistry.cpp [hunk 5]
```diff
@@ -60,43 +58,28 @@ void UnregisterMatchingAttributeAccessInterfaces(F shouldUnregister)
...
+AttributeAccessInterfaceRegistry & AttributeAccessInterfaceRegistry::Instance()
 {
-    UnregisterMatchingAttributeAccessInterfaces(
-        [endpoint = definedEndpoint->endpoint](AttributeAccessInterface * entry) { return entry->MatchesEndpoint(endpoint); });
+    static AttributeAccessInterfaceRegistry instance;
+    return instance;
 }
 
-bool registerAttributeAccessOverride(AttributeAccessInterface * attrOverride)
+void AttributeAccessInterfaceRegistry::UnregisterAllForEndpoint(EmberAfDefinedEndpoint * definedEndpoint)
```
Comment (bzbarsky-apple) [inline src/app/AttributeAccessInterfaceRegistry.cpp:70]: If we are changing this anyway, the argument here should be `EndpointId`.  It was convenient to have it be `EmberAfDefinedEndpoint` back when this all lived in ember stuff, but we don't need that dependency now, right?

Followup is fine for this part.

File: src/app/AttributeAccessInterfaceRegistry.h (hunk 4 above)
Comment (andy31415) [reply src/app/AttributeAccessInterfaceRegistry.h:45]: Updated. Had it lowercase first and `register` was a reserved word.

File: src/app/AttributeAccessInterfaceRegistry.h (hunk 3 above)
Comment (andy31415) [reply src/app/AttributeAccessInterfaceRegistry.h]: Yes, Started renaming and it started lowercase with `register` being a reserved word. Updated.

File: src/app/AttributeAccessInterfaceRegistry.cpp [hunk 6]
```diff
@@ -60,43 +58,28 @@ void UnregisterMatchingAttributeAccessInterfaces(F shouldUnregister)
...
 {
-    gAttributeAccessInterfaceCache.Invalidate();
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/app/AttributeAccessInterfaceRegistry.cpp:77]: I think methods were moved by copy & paster. The invalidate in unregister and register still exist.

File: src/app/InteractionModelEngine.cpp (hunk 1 above)
Comment (andy31415) [reply src/app/InteractionModelEngine.cpp:105]: We are symmetric with AAI: there is no UnregisterAll for AAI, so i figured we would not use it for CHI right now.

We do have a UnregisterAllForEndpoint in endpoint shutdowns.

File: src/app/AttributeAccessInterfaceRegistry.cpp [hunk 7]
```diff
@@ -106,21 +89,44 @@ app::AttributeAccessInterface * GetAttributeAccessOverride(EndpointId endpointId
...
 
         // Did not find AAI registered: mark as definitely not using.
-        gAttributeAccessInterfaceCache.MarkUnused(endpointId, clusterId);
+        mAttributeAccessInterfaceCache.MarkUnused(endpointId, clusterId);
     }
 
     return nullptr;
 }
 
+void AttributeAccessInterfaceRegistry::UnregisterAttributeAccessOverride(AttributeAccessInterface * attrOverride)
```
[the remark this answers is not in the document]
Comment (andy31415) [reply src/app/AttributeAccessInterfaceRegistry.cpp]: Placed them however the copy and paste took them. Updated ordering.

File: src/app/InteractionModelEngine.cpp (hunk 1 above)
Comment (andy31415) [reply src/app/InteractionModelEngine.cpp:105]: I can keep the unregister, however because there is no ownership of this anymore, it is odd. If we run multiple IME in paralle, having any IME shutdown clear the handlers for all the others IME seems bad.

We do not do this for AAI.

Comment (bzbarsky-apple) [reply src/app/InteractionModelEngine.cpp:105]: When we run multiple IME in parallel, each one will have a separate registry is the point.

For AAI, we should have a "remove all on shutdown" as well if we want to allow any sort of non-static registration, but AAI was initially created with static-only things in mind.

Comment (andy31415) [reply src/app/InteractionModelEngine.cpp:105]: I think the question is if this PR is acceptable now or should be deferred until we have a usable DataModel (or DataModel::Provider as the new name) inside the IME.

The reason I am splitting this is to have advance work to decouple IME from all implementations (so they use a provider instead of the IME). My concern is that given that CodeGen metadata is used to list available things, including Generated/Accepted commands, this is still centrally managed and code generated today to some extent.


Review feedback request on path forward:
   - keep the odd clear in here (even if ownership is odd)
   - remove the clear claiming 'things are centrally managed anyway, and since IME does not own the registry, it should not clear it on shutdown'
   
I do not care as much since I imagine medium-term we will have a real data model split anyway which should make things sane (and codegen data model being a singleton can absolutely do the clear of AAI/CHI as it wants).

Comment (bzbarsky-apple) [reply src/app/InteractionModelEngine.cpp:105]: I think this PR is fine in its current general form, but I think we should keep the odd clear in here for now, and later on once we have separate registries for separate stack instances the clear would move into stack instance shutdown.

Comment (andy31415) [reply src/app/InteractionModelEngine.cpp:105]: Added clear back.

File: src/app/AttributeAccessInterfaceRegistry.cpp [hunk 8]
```diff
@@ -60,43 +58,50 @@ void UnregisterMatchingAttributeAccessInterfaces(F shouldUnregister)
...
-    UnregisterMatchingAttributeAccessInterfaces([attrOverride](AttributeAccessInterface * entry) { return entry == attrOverride; });
+    mAttributeAccessInterfaceCache.Invalidate();
+    UnregisterMatchingAttributeAccessInterfaces([attrOverride](AttributeAccessInterface * entry) { return entry == attrOverride; },
+                                                mAttributeAccessOverrides);
 }
 
-void unregisterAllAttributeAccessOverridesForEndpoint(EmberAfDefinedEndpoint * definedEndpoint)
+void AttributeAccessInterfaceRegistry::UnregisterAllForEndpoint(EndpointId endpointId)
 {
     UnregisterMatchingAttributeAccessInterfaces(
```
Comment (bzbarsky-apple) [inline src/app/AttributeAccessInterfaceRegistry.cpp:79]: OK, I see.  This just never invalidated the cache.  But it needs to, no?

Followup is OK for this, since it's pre-existing.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/AttributeAccessInterfaceRegistry.cpp (hunk 8 above)
Comment (andy31415) [reply src/app/AttributeAccessInterfaceRegistry.cpp:80]: Added it to this PR ... may as well fix things.
I think there was a chance we never noticed because we unregister only on EP disabling, at which point the AAI is not usable anyway (and future enabling re-register things that invalidate the cache).

File: scripts/py_matter_idl/matter_idl/test_idl_generator.py [hunk 9]
```diff
@@ -55,8 +55,8 @@ def ReadMatterIdl(repo_path: str) -> str:
         return stream.read()
 
 
-def ParseMatterIdl(repo_path: str, skip_meta: bool) -> Idl:
-    return CreateParser(skip_meta=skip_meta).parse(ReadMatterIdl(repo_path))
+def ParseMatterIdl(repo_path: str, skip_meta: bool, merge_globals: bool) -> Idl:
```
Comment (ksperling-apple) [inline scripts/py_matter_idl/matter_idl/test_idl_generator.py:58]: Is this change related to the title of this PR?

**Review by ksperling-apple — COMMENTED** [review]
If we're changing these APIs anyway, we should consider merging AttributeAccessInterface and CommandHandlerInterface. In most cases cluster servers implement both interfaces, so this would reduce the amount of code required to register a cluster server, as well as memory used by cluster server instances and the registries themselves.

**Comment by andy31415** [comment]
> If we're changing these APIs anyway, we should consider merging AttributeAccessInterface and CommandHandlerInterface. In most cases cluster servers implement both interfaces, so this would reduce the amount of code required to register a cluster server, as well as memory used by cluster server instances and the registries themselves.

I plan to make this change, however for backwards compatibility I am placing after the DM::Provider interface split and usage.

File: scripts/py_matter_idl/matter_idl/test_idl_generator.py (hunk 9 above)
Comment (andy31415) [reply scripts/py_matter_idl/matter_idl/test_idl_generator.py]: No, I think it was a master merge. Updated with a master merge.

**Review by tehampson — APPROVED** [review]

File: src/app/AttributeAccessInterfaceRegistry.h [hunk 10]
```diff
@@ -16,39 +16,49 @@
...
-AttributeAccessInterface * GetAttributeAccessOverride(EndpointId aEndpointId, ClusterId aClusterId);
+    /**
+     *  Get the registered attribute access override. nullptr when attribute access override is not found.
+     */
+    AttributeAccessInterface * Get(EndpointId aEndpointId, ClusterId aClusterId);
+
+    static AttributeAccessInterfaceRegistry & Instance();
+
+private:
+    AttributeAccessInterface * mAttributeAccessOverrides = nullptr;
```
Comment (cecille) [inline src/app/AttributeAccessInterfaceRegistry.h:59]: Is this always intended to be a singleton? If so, would it make sense to hide the constructor in private?

**Review by cecille — APPROVED** [review]

File: src/app/AttributeAccessInterfaceRegistry.h (hunk 10 above)
Comment (andy31415) [reply src/app/AttributeAccessInterfaceRegistry.h:59]: Used as a singleton in the app, however I am unsure if we ever need testing. I have been having a hard time working with code that went too far into protecting from the user (usually with many private things and having friend only access) so I generally default to public for testability unless there are very strong reasons to do otherwise.

Here it probably does not matter as much as we could `clear all` in tests, however I don't see too much of a harm if we allow constructors. Overally globals are not ideal. In the future I would rather expect we want CodegenDataModelProvider to own these if possible (not possible today because all cluster implementations are coupled and assume a global).
